winch/codegen/src/codegen/mod.rs

```diff
@@ -404,10 +408,78 @@ where
 
                 match &ty {
                     I32 | I64 | F32 | F64 => self.masm.store(src.into(), addr, ty.into()),
-                    _ => panic!("Unsupported type {:?}", ty),
+                    Ref(rt) => match rt.heap_type {
+                        WasmHeapType::Func => self.masm.store_ptr(src.into(), addr),
+                        ht => unimplemented!("Support for WasmHeapType: {ht}"),
+                    },
+                    _ => unimplemented!("Support for WasmType {ty}"),
                 }
             });
     }
+
+    /// Emits a series of instructions to lazily initialize a function reference.
+    pub fn emit_lazy_init_funcref(
+        table_data: &TableData,
+        table_index: TableIndex,
+        ptr_type: WasmType,
+        context: &mut CodeGenContext,
+        masm: &mut M,
+        call: &mut FnCall,
+        callee: Reg,
+    ) {
+        let index = context.pop_to_reg(masm, None);
+        let elem_value: Reg = context.any_gpr(masm).into();
+        let base = context.any_gpr(masm);
+        let elem_addr = masm.table_elem_address(index.into(), base, &table_data, context);
+        masm.load_ptr(elem_addr, elem_value);
+
+        let defined = masm.get_label();
+        let cont = masm.get_label();
+
+        // Preemptively move the table element address to the
+        // result register, to avoid conflicts at the control flow merge.
+        let result = call.abi_sig.result.result_reg().unwrap();
+        masm.mov(elem_value.into(), result, ptr_type.into());
+
+        // Push the builtin function arguments to the stack.
+        context
+            .stack
+            .push(TypedReg::new(ptr_type, <M::ABI as ABI>::vmctx_reg()).into());
+        context.stack.push(table_index.as_u32().try_into().unwrap());
+        context.stack.push(index.into());
+
+        masm.branch(
+            CmpKind::Ne,
+            elem_value.into(),
+            elem_value,
+            defined,
+            ptr_type.into(),
+        );
```

Comment:
Yeah I agree this is confusing, and perhaps something that can be changed/improved because thinking a bit more about it, this is probably leaking some target-specific details. Internally `masm.branch` is going to emit a `test` if both registers are the same and if the comparison is zero or not zero (`Eq`, `Ne`). I'll make a note out of this, perhaps we can introduce either a new instruction at the MacroAssembler layer.

Reply:
Would you mind adding a comment explaining the effect of the branch in the meantime? I think adding a `test` function to the macro assembler makes sense, but that doesn't need to block this PR.

Reply:
Done, thanks!